### ISSUE TITLE
Feature Request: Ability to reset an M2 reader to require an update

### ISSUE THREAD

**Issue by jbring17** [stripe/stripe-terminal-android#431]
## Summary
There is logic in our app that handles updating of readers and it is frustrating that we only get to test it once on a reader and then if we need to test it again we need another reader that needs an update.

## SDK version
3.3.1

## Other information
The ability to reset the reader to require an update. Make the factory reset go back a version or something so that you need to update it. We have used the simulated readers, but those don't always give the real world experience of an actual physical device.

**Comment by billfinn-stripe** [comment]
Hi @jbring17 -- thanks for the request. One small follow-up:

> We have used the simulated readers, but those don't always give the real world experience of an actual physical device.

Can you elaborate on some of the differences you are observing?

**Comment by jbring17** [comment]
Hi Bill,

Yeah so when we use the simulated readers to test the updating everything
works as expected. But then when we get to the real readers there are
certain things that can fail / not work such as internet interruption,
bluetooth connection interruption, the reader not having sufficient battery
for the updates. We have run into problems where we tested with the
simulators and everything was working and then we shipped out our app and
customers were having issues while using the real terminals.

- Jerry

**Comment by ashare80** [comment]
Agree. This would be ideal also to simulate a fresh reader unboxing as we just ran into an issue where new readers were unboxed for setup with a client and the new SDK release appears to have broke upgrading from an older firmware version which is a situation we would need a new reader every time to test the onboarding works.

We also had a similar experience where the simulated reader update on past versions of the SDK (haven't tested in it a quite a few versions) did not seem to simulate the case where a reader requires update before connection completes so we needed to get a reader to have an update and figure out all the different cases of what happens on connection or low battery or disconnection.

**Comment by dmatthews-stripe** [comment]
Hi @jbring17 we are taking this on as a feature request, but at this time we do not have a specific timeline as to when we will complete the work.

**Comment by abhiin1947** [comment]
I'm also interested in knowing if this was ever prioritized to work on. We have an M2 which unfortunately calls a bluetooth disconnection error like so

```didFinishInstallingUpdate {"deviceSoftwareVersion": "2.01.00.36-SZZZ_Prod_US_v12-481001", "error": {"code": "READER_ERROR.BLUETOOTH_DISCONNECTED", "message": "Bluetooth unexpectedly disconnected during operation."}, "estimatedUpdateTime": "estimate5To15Minutes", "requiredAt": "1755123852102"}```. It would be great to test this flow and make this more resilient for us.

**Comment by ugochukwu-stripe** [comment]
Hi folks, thanks for your patience on this one.

As of the 5.5.0 release[^1], you can now simulate reader software updates on physical mobile readers in test mode — no need to acquire a new reader each time.

Set `testReaderUpdate` on your `BluetoothConnectionConfiguration` or `UsbConnectionConfiguration` before connecting, using the new `TestReaderUpdate` and `TestReaderUpdateType` classes.
Please refer to the Stripe docs [^2] for more details about scenarios you can simulate.

Note: `SimulatorConfiguration.update` is now deprecated in favor of this new API.

Please give it a try and let us know if it covers your testing needs. Closing this as fixed — feel free to open a new issue if you find any bugs or unexpected behavior.
[^1]: https://github.com/stripe/stripe-terminal-android/blob/master/CHANGELOG.md#550---2026-05-06
[^2]: https://docs.stripe.com/terminal/references/testing?terminal-sdk-platform=android#simulated-reader-updates
